### PR TITLE
Support for template strings used as argument to member expression

### PR DESCRIPTION
**Pull request by aryx**
This fixes https://github.com/returntocorp/semgrep/issues/823

Test plan:
test files included, they now parse correctly without any parse error
exceptions and their AST dump looks ok.

### CODE REVIEW

File: lang_js/analyze/js_to_generic.ml [hunk 1]
```diff
@@ -121,16 +121,19 @@ let special (x, tok) =
...
-      | Some n -> 
-            let n = name n in
-            SR_NeedArgs (fun args ->
-            G.OtherExpr (G.OE_EncapsName,(G.I n)::(args|>List.map(fun e ->G.E e))))
-      )
+  | Encaps has_tag_function -> 
+      if not has_tag_function 
+      then SR_NeedArgs (fun args -> 
+            G.Call (G.IdSpecial (G.ConcatString G.InterpolatedConcat, tok), 
+                    args |> List.map (fun e -> G.Arg e)))
```
Comment (nbrahms) [inline lang_js/analyze/js_to_generic.ml:128]: A nice-to-have here would be to throw the argument to `SR_NeedArgs` in a `with` and give it a nice name (`InterpolatedStringCall`?). Would also allow you to DRY lines 134-136.

File: lang_js/analyze/ast_js.ml [hunk 2]
```diff
@@ -136,7 +136,7 @@ type special =
   | In | Delete 
   | Spread
   | Yield | YieldStar | Await
-  | Encaps of name option (* less: resolve? *)
+  | Encaps of bool (* if true, first arg of apply is the "tag" *)
```
Comment (nbrahms) [inline lang_js/analyze/ast_js.ml:139]: I could also imagine leveraging typing here:

```ocaml
| EncapsCall of encaps
| Encaps
```
where the latter is just a nullary variant.

**Comment by nbrahms** [comment]
I had some minor style nits; please feel free to ignore them.

File: lang_js/analyze/ast_js.ml (hunk 2 above)
Comment (aryx) [reply lang_js/analyze/ast_js.ml:139]: Actually it should be Encaps of expr option, to make the difference between
`foo ${x}` and tag`foo ${x}` but that would require the special type to be mutually recursive with expr, which I don't want. The motivation for special is actually to reduce the size of the cases in the expression, to simplify visitor/mappers and more generally any algorithm working on expressions, but at the same time I don't want to lose too much information and put everything in a call. For me Eval is special, as well as This/Super/... hence this type.
We lose indeed typing information by putting may things in special without any information about the arity of those special functions/values, but that's the tradeoff ... less typing security, but less cases to handle in most algorithms.

File: lang_js/analyze/js_to_generic.ml (hunk 1 above)
Comment (aryx) [reply lang_js/analyze/js_to_generic.ml:128]: I don't understand this comment.

**Pull request merged**